libpromises/evalfunction.c

```diff
@@ -638,28 +638,92 @@ static FnCallResult FnCallHostsWithClass(EvalContext *ctx, ARG_UNUSED const Poli
 
 /*********************************************************************/
 
+/** @brief Convert function call from/to variables to range
+ *
+ *  Swap the two integers in place if the first is bigger
+ *  Check for CF_NOINT, indicating invalid arguments
+ *
+ *  @return Absolute (positive) difference, -1 for error (0 for equal)
+*/
+static int int_range_convert(int *from, int *to)
+{
+    int old_from = *from;
+    int old_to = *to;
+    if (old_from == CF_NOINT || old_to == CF_NOINT)
+    {
+        return -1;
+    }
+    if (old_from == old_to)
+    {
+        return 0;
+    }
+    if (old_from > old_to)
+    {
+        *from = old_to;
+        *to = old_from;
+    }
+    assert(*to > *from);
+    return (*to) - (*from);
+}
+
 static FnCallResult FnCallRandomInt(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const Policy *policy, ARG_UNUSED const FnCall *fp, const Rlist *finalargs)
 {
-    int tmp, range, result;
+    if (finalargs->next == NULL)
+    {
+        return FnFailure();
+    }
 
     int from = IntFromString(RlistScalarValue(finalargs));
     int to = IntFromString(RlistScalarValue(finalargs->next));
 
-    if (from == CF_NOINT || to == CF_NOINT)
+    int range = int_range_convert(&from, &to);
+    if (range == -1)
     {
         return FnFailure();
     }
+    if (range == 0)
+    {
+        return FnReturnF("%d", from);
+    }
 
-    if (from > to)
+    assert(range > 0);
+
+    int result = from + (int) (drand48() * (double) range);
+
+    return FnReturnF("%d", result);
+}
+
+static FnCallResult FnCallHashToInt(ARG_UNUSED EvalContext *ctx, ARG_UNUSED const Policy *policy, ARG_UNUSED const FnCall *fp, const Rlist *finalargs)
+{
+    if (finalargs->next == NULL || finalargs->next->next == NULL)
     {
-        tmp = to;
-        to = from;
-        from = tmp;
+        return FnFailure();
+    }
+    int from = IntFromString(RlistScalarValue(finalargs));
+    int to = IntFromString(RlistScalarValue(finalargs->next));
+
+    int range = int_range_convert(&from, &to);
+    if (range == -1)
+    {
+        return FnFailure();
+    }
+    if (range == 0)
+    {
+        return FnReturnF("%d", from);
     }
+    assert(range > 0);
 
-    range = abs(to - from);
-    result = from + (int) (drand48() * (double) range);
+    const char *inp = RlistScalarValue(finalargs->next->next);
 
+    // Use beginning of SHA checksum as basis:
+    unsigned char digest[EVP_MAX_MD_SIZE + 1];
```

Comment:
Initialize to empty array `{0}`. HashString() is an old function which can fail silently. EDIT: actually it doesn't matter much here.